### DISCUSSION TITLE
make-like file dependencies (aka is just a .PHONY replacement?)

### DISCUSSION THREAD

**Discussion by patmaddox** [casey/just#1382] [General]
It's been really easy to get going with just.

One thing I'm curious about: does it support make-like file dependencies, so it won't run a command if the file is up-to-date?

I think it doesn't, but I want to double check.

In that case, I think I can conclude two things:

1. `just` is essentially a ~cross-platform~ non-make-variant-specific replacement for `.PHONY` tasks (yes I've seen the [idiosyncrasies FAQ](https://github.com/casey/just#what-are-the-idiosyncrasies-of-make-that-just-avoids))
2. If I want to generate files, I should still use `make`.

In other words, it gets me a "best tool for the job." Use `make` for generating files, and `just` for defining commands that I would otherwise shoehorn into `.PHONY` tasks. Does that sound right?

**Comment by casey** [comment]
You're correct, `just` does not support make-like file dependencies. Although I've thought of a syntax for adding it that I don't mind. Recipes with a name that's a string, instead of an identifier, would be file dependencies:

```make
"a.out": "main.c"
  cc main.c
```

There's a lot more to figure out and implement, so it'll be a while before it's ready. In particular, Make has a lot of gotchas, so I want to make sure I avoid those.

> In other words, it gets me a "best tool for the job." Use make for generating files, and just for defining commands that I would otherwise shoehorn into .PHONY tasks. Does that sound right?

Yup, that's correct. And you could also look at something like `ninja` as an alternative to `make`.

**Reply by patmaddox** [reply]
Cool, thanks, that makes sense.

> you could also look at something like `ninja` as an alternative to `make`

Interesting, I hadn't come across this before. I've read through the manual a couple times, and some of the associated projects, still trying to wrap my head around it.

It seems like it's useful in things like C++ where you just have a compiler, but nothing to traverse all the source files. Whereas Rust has `cargo`, and in Elixir we have `mix`, so there's probably no point to using `ninja` there. Then again, that's back to my original point - that `just` is a good wrapper for `mix` _and_ `make` (or `ninja`), or any build system.

The thing I may have to careful of is mixing these things up... because I can see how it would be useful for `just` to call the build systems - but also for build systems to call `just` (e.g. there's probably a useful `just` command to generate the `ninja.build` file). So there may be some aspect of layering here - a `build.justfile` with commands intended to be called from the build system, and a `justfile` that provides the top-level commands, which may include calling into the build system.

**Comment by danielb2** [comment]
I've been wondering about this too, so thanks for bringing it up pat and thanks for responding casey. I really make use of the dependencies for building stuff, but I think just is awesome. So I've kept using Makfiles for most  things, and justfiles for the things that are just jobs.
